### PR TITLE
Added Shopping cart and endpoints to interact with it

### PR DESCRIPTION
**Pull request by IhorTrokhymchuk**

### CODE REVIEW

File: src/main/java/com/example/store/dto/cartitem/CartItemRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.example.store.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Min(1)
```
Comment (okuzan) [inline src/main/java/com/example/store/dto/cartitem/CartItemRequestDto.java]: @ Positive

File: src/main/java/com/example/store/model/ShoppingCart.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany(cascade = CascadeType.ALL)
+    @JoinTable(name = "shopping_carts_cart_items",
+            joinColumns = @JoinColumn(name = "shopping_cart_id"),
+            inverseJoinColumns = @JoinColumn(name = "cart_items_id")
+    )
+    private Set<CartItem> cartItems = new HashSet<>();
+    @Column(name = "is_deleted",nullable = false)
```
Comment (okuzan) [inline src/main/java/com/example/store/model/ShoppingCart.java]:
```suggestion
    @Column(name = "is_deleted", nullable = false)
```

File: src/main/java/com/example/store/repository/user/UserRepository.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 public interface UserRepository extends JpaRepository<User, Long> {
     Optional<User> findUserByEmail(String email);
 
-    @Query("SELECT u FROM User u INNER JOIN FETCH u.roles r WHERE u.email = :email")
+    @Query("SELECT u FROM User u JOIN FETCH u.roles WHERE u.email = :email")
     Optional<User> findUserByEmailWithRoles(String email);
+
```
Comment (okuzan) [inline src/main/java/com/example/store/repository/user/UserRepository.java]:
```suggestion
```

File: src/main/java/com/example/store/service/ShoppingCartService.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import com.example.store.model.User;
+
+public interface ShoppingCartService {
+    ShoppingCartDto registerNewShoppingCart(User user);
+
+    ShoppingCartDto addItemToShoppingCart(CartItemRequestDto requestDto, String email);
+
+    ShoppingCartDto findShoppingCart(String email);
+}
+
```
Comment (okuzan) [inline src/main/java/com/example/store/service/ShoppingCartService.java]:
```suggestion
```

File: src/main/java/com/example/store/service/impl/CartItemServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    public CartItem save(CartItemRequestDto requestDto, ShoppingCart shoppingCart) {
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(cartItem -> cartItem.getBook().getId().equals(requestDto.getBookId()))
+                .findFirst();
+
+        if (existingCartItem.isPresent()) {
+            CartItem cartItemToUpdate = existingCartItem.get();
+            cartItemToUpdate.setQuantity(requestDto.getQuantity());
+            return cartItemRepository.save(cartItemToUpdate);
+        } else {
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/CartItemServiceImpl.java:37]: else after return is useless

File: src/main/java/com/example/store/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public ShoppingCartDto registerNewShoppingCart(User user) {
+        return shoppingCartMapper.toDto(createShoppingCart(user));
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addItemToShoppingCart(CartItemRequestDto requestDto, String email) {
+        ShoppingCart shoppingCart = getShoppingCart(email);
+        shoppingCart.getCartItems().add(cartItemService.save(requestDto, shoppingCart));
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/ShoppingCartServiceImpl.java]: i'm not sure what you are doing here, but it looks inefficient. Are you sure this is not redundant?

File: src/main/java/com/example/store/service/impl/CartItemServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/CartItemServiceImpl.java:27]: better add transactional

File: src/main/java/com/example/store/service/impl/CartItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (existingCartItem.isPresent()) {
+            CartItem cartItemToUpdate = existingCartItem.get();
+            cartItemToUpdate.setQuantity(requestDto.getQuantity());
+            return cartItemRepository.save(cartItemToUpdate);
+        } else {
+            CartItem cartItem = cartItemMapper.toModel(requestDto);
+            cartItem.setBook(bookRepository.findById(requestDto.getBookId())
+                    .orElseThrow(() -> new EntityNotFoundException("Can't find book with id: "
+                            + requestDto.getBookId())));
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/CartItemServiceImpl.java]: extract this to a local variable for better readability

File: src/main/java/com/example/store/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    private ShoppingCart getShoppingCart(String email) {
+        Optional<ShoppingCart> existingShoppingCart =
+                shoppingCartRepository.findShoppingCartAndCartItemByUserEmail(email);
+
+        return existingShoppingCart.orElseGet(
+                () -> createShoppingCart(userRepository.findUserByEmail(email)
+                        .orElseThrow(
+                                () -> new EntityNotFoundException("Cant find user with email: "
+                                        + email)
+        )));
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/ShoppingCartServiceImpl.java]: not a mistake, but a bit weird formatting, is that google-java-format plugin? it works bad with lambdas

File: src/main/java/com/example/store/service/impl/CartItemServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public CartItem save(CartItemRequestDto requestDto, ShoppingCart shoppingCart) {
```
Comment (okuzan) [inline src/main/java/com/example/store/service/impl/CartItemServiceImpl.java]:
```suggestion
    public CartItem addOrUpdateCartItem(CartItemRequestDto requestDto, ShoppingCart shoppingCart) {
```
maybe smth like this to better reflect it's logic

File: src/main/java/com/example/store/controller/ShoppingCartController.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @GetMapping
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "Get all categories",
+            description = "Get a page of all available categories and can use sort")
+    public ShoppingCartDto getAll(Authentication authentication, Pageable pageable) {
+        return shoppingCartService.findShoppingCart(authentication.getName());
```
Comment (okuzan) [inline src/main/java/com/example/store/controller/ShoppingCartController.java:39]: imho it's more usual to work with Ids

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/store/controller/ShoppingCartController.java (hunk 11 above)
Comment (Vasul-Pankiv) [reply src/main/java/com/example/store/controller/ShoppingCartController.java:39]: agree, using id is much better and faster

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**